src/dune_engine/source_tree.ml

```diff
@@ -476,6 +476,36 @@ end = struct
       (visited, init)
   end
 
+  let ensure_dune_project_file_exists =
+    let memo =
+      Memo.create "ensure-dune-project-file-exists"
+        ~input:(module Dune_project)
+        (fun project ->
+          let open Memo.Build.O in
+          let+ exists =
+            Dune_project.file project |> Path.source |> Fs_memo.file_exists
+          in
+          if not exists then
+            let is_error =
+              match !Clflags.dune_project_file with
+              | Error -> true
+              | Warn -> false
+              | Ignore -> assert false
```

Comment:
You can pass `is_error` as an input to this Memo computation to avoid `assert false`.